### ISSUE TITLE
[Proposal] Automatic foreign key casting for ints

### ISSUE THREAD

**Issue by timacdonald** [laravel/internals#570]
What are your thoughts on having foreign keys automatically cast to integers. I like the idea that when I access a foreign key it is an integer, if not for any other reason than it IS an integer, but hate that I have to do it for all my foreign keys if I want this functionality.

I know in our dynamic and forgiving PHP land we can work with integers as strings, but I still like this approach (in my own code).

I've put [together a Gist](https://gist.github.com/timacdonald/3542d9544e7eb13ba3fce86867a11585) of how I do this in my current setup, obviously there is probably a better way if it was actually integrated into the core.

This is just me wanting my snozberries to *taste like snozberries*.

Basically when checking for casting, the model checks for all foreign keys and cast them appropriately.

Thoughts? Crazy stupid? Meh?

**Comment by Garbee** [comment]
FKs don't need to be int. UUID for example.

**Comment by timacdonald** [comment]
Totally agree. But in the case where they are this is nice (I think). Wouldn't need to be on by default, perhaps a "castsForeignKeys" flag on the model or something?

My Gist is not a core implementation example, just something I threw together for my own projects.

**Comment by Dylan-DPC** [comment]
Can you provide an example ?

**Comment by Garbee** [comment]
How would this interact though on a model that has multiple FK's to different tables which are either auto-incrementing (or some numeric type) **or** UUID (or some string type)?

I don't see where anything is currently broken or made difficult by the keys not being ints (which also may simply be due to the database driver you're using.) It's a trivial thing to handle when you need it and otherwise it doesn't get in anythings way.

**Comment by timacdonald** [comment]
@Garbee you could very well be onto something regarding it being the SQL driver producing strings instead of ints and this is really just a nicety.

@Dylan-DPC sure, imagine this is in the Eloquent model, I've modified my version to suit what I see in the model (no constants) and also just moved the casting type into a string in the method cause really, your only probably going to go to an int.

I've also added a `$castsForeignKeys` flag and set it to false by default, but if its in there, maybe it should be automatic?

```php
class Model
{
    protected $castsForeignKeys = false;

    protected $foreignKeyCastingExceptions = [];

    protected $foreignKeyRegex = '/.+_id$/';

    protected $foreignKeyCastsAdded = false;

    //

    public function getCasts()
    {
        if (!$this->foreignKeyCastsAdded) {
            $this->addForeignKeyCasts();
        }

        // the rest of this method
    }
    
    protected function addForeignKeyCasts()
    {
        foreach ($this->attributes as $key => $value) {
            if ($this->shouldCastForeignKey($key)) {
                $this->addForiegnKeyCast($key);
            }
        }

        $this->foreignKeyCastsAdded = true;
    }

    protected function addForiegnKeyCast($key)
    {
        if (!in_array($key, $this->casts)) {
            $this->casts[$key] = 'integer';
        }
    }

    protected function shouldCastForeignKey($key)
    {
        return $this->isForeignKey($key)
           && !$this->isForeignKeyCastingException($key);
    }

    protected function isForeignKeyCastingException($key)
    {
        return in_array($key, $this->foreignKeyCastingExceptions);
    }

    protected function isForeignKey($key)
    {
        return preg_match($this->foreignKeyRegex, $key);
    }
}
```

Then we have a (classic) `Post`. Our post has 3 foreign keys, a UUID and 2 ints.

```php
class Post extends Model
{
    $castsForeignKeys = true;

    $foreignKeyCastingExceptions = [
        'my_uuid_foreign_key_id'
    ];
}

$post = Post::first();

// string
$post->my_uuid_foreign_key_id

// int
$post->my_int_foreign_key_1_id

// int
$post->my_int_foreign_key_2_id
```

**Comment by timacdonald** [comment]
Again, this is just a modified version of what I'm using so not going to be the best way to do it, but an example of a possible path to implementation code.

**Comment by timacdonald** [comment]
After doing this a little, I really am not find any benefit.
